### PR TITLE
[MLIR] Generalize expand_shape to take shape as explicit input

### PR DESCRIPTION
**Pull request by Shukla-Gaurav**
This patch generalizes tensor.expand_shape and memref.expand_shape to consume the output shape as a list of SSA values.  This enables us to implement generic reshape operations with dynamic shapes using collapse_shape/expand_shape pairs.

The output_shape input to expand_shape follows the static/dynamic representation that's also used in `tensor.extract_slice`.

Differential Revision: https://reviews.llvm.org/D140821

### CODE REVIEW

File: mlir/include/mlir/Dialect/Utils/StaticValueUtils.h
Comment (MaheshRavishankar) [inline mlir/include/mlir/Dialect/Utils/StaticValueUtils.h:112]: I am not sure this is needed.

File: mlir/lib/Conversion/TosaToTensor/TosaToTensor.cpp
Comment (MaheshRavishankar) [inline mlir/lib/Conversion/TosaToTensor/TosaToTensor.cpp:205]: We could just use a helper build method to do this automatically instead of adding it at every call site.

File: mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:458]: Nit: add `{` `}` for multiline statements.

File: mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorRewriting.cpp
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorRewriting.cpp:457]: Same here. Use `{` `}` for multi-line statements.

File: mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:64]: We want to avoid creating `arith.divui` ops directly. Use `affine.apply` operations instead. Those fold better.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Ok, I think this is not looking too bad. Adding the helper build method here would make the diff much smaller I think.

**Comment by MaheshRavishankar** [comment]
Hey, I left a few comments. Let me know when this is ready for review.

File: mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h
Comment (Shukla-Gaurav) [reply mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h:196]: Yeah, we can get it by ReshapeOpTy itself.
I guess this extra template parameter maybe added to simplify the check as this utility can be used by reshape ops of many dialects(currently used by memref and tensor dialects). @ramiro050 may have more context on this.

File: mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h
Comment (Shukla-Gaurav) [reply mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h:49]: This method only returns success or failure, the output_shape is pass by reference.

File: mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorRewriting.cpp
Comment (Shukla-Gaurav) [reply mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorRewriting.cpp:823]: Did you mean some other if stmt, as this one already have.

File: mlir/lib/Conversion/TosaToTensor/TosaToTensor.cpp
Comment (Shukla-Gaurav) [reply mlir/lib/Conversion/TosaToTensor/TosaToTensor.cpp]: I have added a helper build method which can infer the output shape. But in the cases where we can not infer output shapes, it's assert vs failure. With the helper build method, in case the output shapes can not be inferred, it breaks with assert(let me know if there is better way?). 

If the helper build method seems fine, I will update other parts of the code to use this build method.
cc: @ramiro050

File: mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp
Comment (Shukla-Gaurav) [reply mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:64]: working on this.

File: mlir/include/mlir/Dialect/Utils/StaticValueUtils.h
Comment (Shukla-Gaurav) [reply mlir/include/mlir/Dialect/Utils/StaticValueUtils.h:127]: Yeah, it seems this patch does not necessarily require this method. If the helper build method seems fine(mentioned in the next comment), I will address this comment together with the next one.

File: mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h
Comment (ramiro050) [reply mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h:196]: See: https://reviews.llvm.org/D140821#inline-1369788

Comment (ramiro050) [reply mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h:196]: From Sanjoy's comment in link above:

> Unless I'm missing some C++ trick, that requires to add a dependency from this header to either the memref or the tensor dialect. I assumed that was not OK, but please let me know if it is.

File: mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td
Comment (MaheshRavishankar) [inline mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td:1075]: Please take these out of `asserts` . This will be ignored during release builds. SO
```
auto status = inferOutputShape(...)
(void) status
assert(succeeded(status) && ...);
```

File: mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp
Comment (MaheshRavishankar) [inline mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp:555]: With the new build method you should be able to fold this into the `create` method? Here and everywhere else?

File: mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:79]: Just return `failure` here.

Also it seems like there is really no `failure` returning happening in this method. One option is to make this function return `void` (in which case the `assert` here is OK). On balance, maybe better to just keep the `LogicalResult` return type and return `failure()`.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Next rounds of comments. Thanks!

File: mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp
Comment (Shukla-Gaurav) [reply mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:64]: Thanks, updated with affine.apply.

File: mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp
Comment (Shukla-Gaurav) [reply mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp]: Updated at all places, thanks!

**Comment by Shukla-Gaurav** [comment]
@MaheshRavishankar Thanks for reviewing, have addressed all the comments, and is ready for re-review.
@ramiro050 @sanjoy Thanks for implementing this, it was needed for a long time. As I am continuing this, would love to have your review on this.

**Review by MaheshRavishankar — COMMENTED** [review]
ok, this looks good to me. Lets test this with IREE. Ill let you know how to do that.

**Review by ramiro050 — APPROVED** [review]
Thanks for getting this across the finish line! LGTM

File: mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp
Comment (newling) [inline mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:92]: nit:

outputShapeValues has the same size as the input's rank. I think it could be initialized at this size, and populated by the reassociation group index. i.e. 

 `outputShapeValues[inputIndex] = dynamicDimSize;` instead of `outputShapeValues[index] = dynamicDimSize;`

File: mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp
Comment (newling) [inline mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:240]: aI'd like to get a better understanding of this PR from a high-level. Is it now permitted to have multiple dynamic dimensions in an association? 

Currently this is what happens when you try to expand to multiple dynamic dimensions:

```
foo.mlir:3:8: error: 'tensor.expand_shape' op invalid to have a single dimension (0) expanded into multiple dynamic dims (0,1)
  %0 = tensor.expand_shape %arg0 [[0, 1], [2]] : tensor<?x?xf32> into tensor<?x?x?xf32>
  ```
  
 Is the idea that with this PR this is allowed, but only as long as the dynamic dimensions have values associated to them?
  
  I guess one high-level question is, does this make expand_shape more general purpose, and does that mean that the optimisations that can be applied to it are less powerful?
  
  Any additional high-level info would be welcome :-)

Comment (MaheshRavishankar) [reply mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:240]: I think this is generally increasing the expressive power of the expand_shape op. You are right, it allows you to represent "how" a dynamic dimension is split into multiple dynamic dimensions. Most transformations AFAIK should not care about this split. The load-bearing information is know how the dimensions are split, i.e. the information being carried by reassociation indices... 
I wouldnt say its making expand_shape more general purpose. You are still splitting a dimension, into multiple, its just that those multiple dimensions can now be dynamic. The information about the final output shape is now carried in the op, making the split deterministic. Not sure if that makes sense, but I can explain on a call if that help.s

**Comment by Shukla-Gaurav** [comment]
Currently debugging following issue in this PR:(this issue arises while testing the patch in IREE https://github.com/openxla/iree/pull/16972 )

**add_static.mlir**
```
func.func @main(%input : tensor<1x8x8x3xf32>) -> (tensor<1x8x8x3xf32>) {
  %1 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>, affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>, affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%input, %input : tensor<1x8x8x3xf32>, tensor<1x8x8x3xf32>) outs(%input : tensor<1x8x8x3xf32>) {
  ^bb0(%in: f32, %in_0: f32, %out: f32):
    %3 = arith.addf %in, %in_0 : f32
    linalg.yield %3 : f32
  } -> tensor<1x8x8x3xf32>
  return %1 : tensor<1x8x8x3xf32>
}
```
`/bin/mlir-opt  --test-linalg-drop-unit-dims add_static.mlir`
[stack_dump.txt](https://github.com/llvm/llvm-project/files/14874018/stack_dump.txt)

cc: @MaheshRavishankar @pashu123

**Comment by MaheshRavishankar** [comment]
> Currently debugging following issue in this PR:(this issue arises while testing the patch in IREE [openxla/iree#16972](https://github.com/openxla/iree/pull/16972) )
> 
> **add_static.mlir**
> 
> ```
> func.func @main(%input : tensor<1x8x8x3xf32>) -> (tensor<1x8x8x3xf32>) {
>   %1 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>, affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>, affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%input, %input : tensor<1x8x8x3xf32>, tensor<1x8x8x3xf32>) outs(%input : tensor<1x8x8x3xf32>) {
>   ^bb0(%in: f32, %in_0: f32, %out: f32):
>     %3 = arith.addf %in, %in_0 : f32
>     linalg.yield %3 : f32
>   } -> tensor<1x8x8x3xf32>
>   return %1 : tensor<1x8x8x3xf32>
> }
> ```
> 
> **Command** `/bin/mlir-opt --test-linalg-drop-unit-dims add_static.mlir` [stack_dump.txt](https://github.com/llvm/llvm-project/files/14874018/stack_dump.txt)
> 
> cc: @MaheshRavishankar @pashu123

Looking at the stack error seems like it isnt calling the "right" version of the build....

File: mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorRewriting.cpp
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorRewriting.cpp:956]: can this be hidden inside a templated "create" helper method, so we don't get such a huge if-then-else block?

File: mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp
Comment (Shukla-Gaurav) [reply mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp]: Thanks for pointing this, updated the code :)

File: mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp
Comment (ScottTodd) [inline mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:13]: Is this dependency from `Dialect/Utils/` onto the Arith dialect acceptable, or should this new `inferExpandShapeOutputShape()` function be moved to somewhere else (`Dialect/Arith/Utils`? Or memref _and_ tensor where the code is used, with some code duplication?)

**Comment by Shukla-Gaurav** [comment]
@MaheshRavishankar Could you please re-review it? There is still one failure in reshape_fusion(as discussed), meanwhile I can fix that. Thanks :)

**Comment by aartbik** [comment]
I added a comment that has not been addressed yet?

File: mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td
Comment (MaheshRavishankar) [inline mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td:1136]: Can we move the implementation of this into c++ file `Tensor/IR/TensorOps.cpp`. Not good practice to have implementation in the tablegen file, especially when its doing non-trivial things.

File: mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td
Comment (MaheshRavishankar) [inline mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td:1153]: While here, maybe move these to c++ file as well.

File: mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td
Comment (MaheshRavishankar) [inline mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td:1139]: total nit: but if you want to do this, the canonical way of doing this would be to return
```
FailureOr<SmallVector<OpFoldResult>> inferOutputShape(...)
```

That way you can return a mix of values and int64_t as a single vector and not have to return by reference.

File: mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td
Comment (MaheshRavishankar) [inline mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td:1642]: I saw this file after the `TensorOps.td` file. So same comments from there apply here too.. Move these to C++ file

File: mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td
Comment (MaheshRavishankar) [inline mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td:1645]: Nit : Echoing comment from earlier. This would be much more readable with `FailureOr<SmallVector<OpFoldResul>>` . Fine to do it later too.

File: mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:591]: nit: unnecessary empty line

File: mlir/lib/Dialect/Linalg/Transforms/DataLayoutPropagation.cpp
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/Linalg/Transforms/DataLayoutPropagation.cpp:760]: Why did you have to change this? This should have always been a ranked tensor type. If so then something is off with the pack op. Fine to leave as is, just noticing.

File: mlir/lib/Dialect/Linalg/Transforms/DropUnitDims.cpp
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/Linalg/Transforms/DropUnitDims.cpp:256]: This is not returning a failure yet? So then this doesnt need to be a `FailureOr`

File: mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp:821]: If you change the `isFusableMethod` you can just remove this part I think.

File: mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h
Comment (MaheshRavishankar) [inline mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h:50]: I see later that you are using the same method for `MemRefType` and `RankedTensorType`. Maybe make the `expandedType` be `ShapedType` that way you dont need to create a tensor type just to call this method.

File: mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:2245]: Note: This triggered the suggestion of using `ShapedType` in `inferExpandShapeOutputShape` to avoid having to do this.

File: mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:297]: Nice cleanup!

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Thanks! Left a few mostly minor comments. Nothing that I see as blocker for landing.

File: mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td
Comment (bjacob) [inline mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td:1596]: I think this example is missing `output_shape` ?

File: mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td
Comment (Shukla-Gaurav) [reply mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td]: Moved the implementation to TensorOps.cpp for the first two builders as they contain non-trivial code. Thanks :)

File: mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp
Comment (Shukla-Gaurav) [reply mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:293]: Thanks :)

File: mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorRewriting.cpp
Comment (Shukla-Gaurav) [reply mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorRewriting.cpp:956]: There are issues with templated helper. There are few more things which has to be handled for this patch like fusion, I can raise a PR with this and other issues fixed. It has lots of changes, so trying to not hold it for longer. Thanks :)

File: mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp
Comment (Shukla-Gaurav) [reply mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:13]: I tried moving it to tensor/utils, but there seem to be circular dependencies. I can handle this in the next PR. Can we go ahead for this patch as it is?

File: mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorRewriting.cpp
Comment (bjacob) [reply mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorRewriting.cpp:956]: +1, i have a bunch of things I want to do that depend on this, thanks for deferring improvements to follow-up PRs!

I'm also available to help with C++ templates if needed.

File: mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp
Comment (bjacob) [reply mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:13]: +1, see my other comment, thanks for deferring further improvements to follow-up PRs!

File: mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp:810]: Can we move this check into `isFusableWithReshapeByDimExpansion`?

**Review by MaheshRavishankar — APPROVED** [review]
Ok, looks ogod to me. Left a single comment but it is not a blocker and is OK to do it after this PR as well since that part needs to be fixed after this PR lands.

**Comment by bjacob** [comment]
> Ok, looks ogod to me. Left a single comment but it is not a blocker and is OK to do it after this PR as well since that part needs to be fixed after this PR lands.

yes please please do as follow-up and not lose the progress on (very slow) CI here :-)

File: mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp
Comment (Shukla-Gaurav) [reply mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp:810]: Yes, this needs to be fixed anyway and it was easier to place this check here as we have expansionInfo.

**Comment by bjacob** [comment]
Windows CI failed with

> fatal error C1060: compiler is out of heap space

Probably unrelated to this PR, but also not completely impossible that something about this PR increases compiler memory usage and pushes it over a threshold, so we need to get a green CI run before merging. Unfortunately I don't see a "retrigger" button here. Can you push an empty commit to this PR to cause a retrigger maybe?

**Comment by joker-eph** [comment]
I think this broke the bot: https://lab.llvm.org/buildbot/#/builders/61/builds/57064

**Comment by joker-eph** [comment]
Reverted in https://github.com/llvm/llvm-project/pull/89540

**Pull request merged**
